Updated smirky sampler with commandline tool

bannanc opened the pull request:
This pull request has a lot of changes including updates to setup.py to install the command line tool `smirky` (addressed in issue #128 ). 

While it does not actually create the proposal engine addressed in issue #110 the create_new_environment function in sampler_smirky.py is significantly more modular than before so much of the code should be reusable. 

The sampler attempts to address detailed balance, only removing things it can create in one move as discussed in issue #99, but this will likely remain a concern any time we make changes to the way moves are made. 

This is now a separate tool from smarty with it's only driver (cli_smirky.py) and is callable with smirky --options

I haven't updated anything in the main README, but there are two existing examples at [examples/smirky/default_angle](https://github.com/open-forcefield-group/smarty/tree/environment/examples/smirky/default_angle) and [examples/smirky/dummy_torsion_sampler](https://github.com/open-forcefield-group/smarty/tree/environment/examples/smirky/dummy_torsion_sampler) 

It should be noted that while testing I created and destroyed a few different log files causing the number of deletions to be a bit ridiculous. 

There are a lot of changes to environment.py and sampler_smirky.py. I updated the tests and made the new examples, but it's possible there may be problem with older examples using these, i.e. @cbayly13 's ipython notebooks.

davidlmobley committed:
Partial work towards updating _Topology

bannanc committed:
structural changes to environments

davidlmobley committed:
Retrieve version info and aromaticity model from FFXML and store in ForceField.

davidlmobley committed:
Error checking (consistent version and aromaticity model?) in parsed ffxml files.

davidlmobley committed:
Add aromaticity model support to SMIRKS matching function.

davidlmobley committed:
Attempt to use aromaticity model if specified.

davidlmobley committed:
Merge branch 'master' of https://github.com/open-forcefield-group/smarty into issue125

davidlmobley committed:
Adding forcefield which asks for a specific aromaticity model.

bannanc committed:
intermediate updates to smirky sampler and environments)

davidlmobley committed:
More work on implementing bond orders.

bannanc committed:
working command line smirky with many changes to environment and sampler

bannanc committed:
Merge branch 'master' of github.com:open-forcefield-group/smarty into environment

pull origin master

bannanc committed:
fixing example files

bannanc commented:
@davidlmobley when you have time could you take a look?
@cbayly13 you will probably want to see what this is doing so far and maybe comment on the odds used for move making inside the sampler. 

I'm going to make a more complicated example tomorrow with specified decorator/odds files that will hopefully make the moves smarter. 

I do calculated the probability of making each move when a child is created, but then I don't do anything with it. I can print it to the log file if we want. 

On the topic of log files - they are the same format as the printed comments in the original sampler, but looking at the torsion file is painful because most lines don't fit on a computer screen so I will likely re-think some of that structure

bannanc commented:
Note - (mostly to myself) I need to add more tests for environment objects for the new methods, planning to do that tomorrow.

davidlmobley commented:
I'll review relatively soon. I'm not sure this is highest priority though so maybe we can talk about that very briefly today. Thanks.

bannanc committed:
updating tests

bannanc committed:
fixing bug in smirky sampler test

bannanc committed:
successful test on Bonds and parent tree added

davidlmobley committed:
Add storage of bond orders in _Topology.

davidlmobley committed:
Merge branch 'master' of https://github.com/open-forcefield-group/smarty into issue53

davidlmobley committed:
Merge branch 'master' of https://github.com/open-forcefield-group/smarty into issue53

bannanc committed:
more examples with custom decorators

bannanc commented:
Well, I removed the WIP because I thought it was ready to merge, but it looks like tests are failing because there is no module called "smarty"? 

I'll try to look into it, but if any of you know why that might be happening please let me know

davidlmobley commented:
It looks like travis is getting stuck "fetching packages". I don't know why this happens, but I've seen it before. @jchodera , do you know why this happens and what we can do about it?

davidlmobley committed:
First pass at adding a sample ffxml with partial bond orders. Still needs torsions.

davidlmobley committed:
First draft of very rough ffxml file.

davidlmobley committed:
Updating usage info in example ffxml.

bannanc commented:
Thanks David, the good news is the tests are passing on the computer so I don't think I've broken anything so this should be good once we address that issue.

davidlmobley committed:
Update assign charges to also store Wiberg bond order.

davidlmobley committed:
Update _assignPartialCharges so it can be used solely to update bond orders if desired.

davidlmobley committed:
Switch initializeBondOrders to updateBondOrders and implement Wiberg bond order assignment.

davidlmobley commented:
@bannanc -- @jchodera just fixed the tests in a separate PR. (Thanks, John!)

davidlmobley committed:
Update createSystem to obtain partial bond orders and store in topology.

davidlmobley committed:
Merge branch 'master' of https://github.com/open-forcefield-group/smarty into issue53

davidlmobley committed:
First pass at adding partial bond orders.

davidlmobley committed:
First pass at adding partial bond orders.

davidlmobley committed:
Fix notation in ffxml

davidlmobley committed:
Get partial bond orders working for bonds in benzene.

davidlmobley committed:
Cleanup of debug info

davidlmobley committed:
Adding IPython notebook applying bond parameters using partial bond orders to benzene.

bannanc committed:
Merge branch 'master' of github.com:open-forcefield-group/smarty into environment

pull origin master

bannanc committed:
update parse_odds_file function to be python3 compatible

bannanc commented:
Thanks @davidlmobley I had a small python3 compatibility I needed to change, but otherwise this seems to be working well. 

The sampler found all the AlkEtOH bonds in about 600 steps,[ here ](https://github.com/open-forcefield-group/smarty/tree/environment/examples/smirky/custom_Bond). I'm still playing around with odds for angles and torsions, which may end up including built in odds within sampler_smirky.py

Also, for @camizanette , in order to get the commandline tool working, I made some changes to setup.py and moved cli.py to cli_smarty.py, but didn't actually change anything inside cli.py so it shouldn't affect anything you're working on, but the script names on your branch may need to be changed to match the two tool format.

davidlmobley committed:
Had accidentally removed parameter ID from bond parsing

davidlmobley committed:
Adding test of partial bond order code.

davidlmobley committed:
Remove extraneous print

davidlmobley committed:
Switch to beta of OpenEye toolkits

davidlmobley committed:
Update comments in .travis.yml

davidlmobley committed:
Correct beta url

davidlmobley committed:
Further fix for beta

davidlmobley committed:
Merge pull request #134 from open-forcefield-group/issue53

Implement partial bond orders and handle them for bond parameters; switch to October beta OE toolkits

bannanc commented:
@davidlmobley is there something else you think this needs before it can be merged?

davidlmobley commented:
Well, someone should probably review it since it's a rather major change -- I was going to as soon as I get past the deadlines. Or you could recruit someone else.

Also you should update to master.

bannanc commented:
It probably makes sense to wait and have you review when you have time, with Camila out of town and Chris leaving I don't think there's a good alternative. I'll pull from master again.

davidlmobley committed:
Merge MDL aromaticity changes with bond order changes

davidlmobley committed:
Add testing of aromaticity model support.

davidlmobley committed:
Merge pull request #127 from open-forcefield-group/issue125

Add aromaticity model support as per 125

bannanc committed:
Merge branch 'environment' of github.com:open-forcefield-group/smarty into environment

environment branch

davidlmobley commented:
@bannanc - I'm going to review now. But can you also bring this up to date with master?

davidlmobley commented:
First place I went, @bannanc , was to your smirky custom_Bond example, which does not run for me. I get an error like:
```
Attempting to parse file 'atom_OR_bases.smarts'
Attempting to parse file 'atom_OR_bases.smarts'
Traceback (most recent call last):
  File "/Users/dmobley/anaconda/bin/smirky", line 9, in <module>
    load_entry_point('smarty==0.1.2', 'console_scripts', 'smirky')()
  File "/Users/dmobley/anaconda/lib/python2.7/site-packages/smarty-0.1.2-py2.7.egg/smarty/cli_smirky.py", line 138, in main
    atom_OR_bases = smarty.utils.parse_odds_file(option.atom_OR_bases, verbose)
  File "/Users/dmobley/anaconda/lib/python2.7/site-packages/smarty-0.1.2-py2.7.egg/smarty/utils.py", line 226, in parse_odds_file
    for l in input_lines:
UnboundLocalError: local variable 'input_lines' referenced before assignment
```

It looks like some sort of issue with exception handling -- it seems like it's not finding the file .

I think it's a typo in the relevant function in utils.py, where it has `input_ddlines = ifs.readlines()` where it should be `input_lines = ifs.readlines()`. **UPDATE** yes, fixing that makes it run.

So, one small bug to fix so far.

davidlmobley commented:
I also get an error when it finishes, @bannanc :

```Traceback (most recent call last):
  File "/Users/dmobley/anaconda/bin/smirky", line 9, in <module>
    load_entry_point('smarty==0.1.2', 'console_scripts', 'smirky')()
  File "/Users/dmobley/anaconda/lib/python2.7/site-packages/smarty-0.1.2-py2.7.egg/smarty/cli_smirky.py", line 177, in main
    if verbose: print("Final score was %.3f%%" % frac_found*100.)
TypeError: can't multiply sequence by non-int of type 'float'
```

bannanc commented:
@davidlmobley I'll fix the bug in utils and take a look at the second error

bannanc committed:
fix ddlines bug in smarty/utils.py

bannanc commented:
@davidlmobley I tried pulling from master, it said this is up to date, let me know if you think that is wrong for some reason. 
I fixed the input_ddlines bug in smarty/utils.py, I'm surprised that didn't come up earlier for me, but it's fixed now. 

>> ```Traceback (most recent call last):
File "/Users/dmobley/anaconda/bin/smirky", line 9, in 
load_entry_point('smarty==0.1.2', 'console_scripts', 'smirky')()
File "/Users/dmobley/anaconda/lib/python2.7/site-packages/smarty-0.1.2-py2.7.egg/smarty/cli_smirky.py", line 177, in main
if verbose: print("Final score was %.3f%%" % frac_found*100.)
TypeError: can't multiply sequence by non-int of type 'float'

I cannot figure out what is going on with this error, the frac_found should be a float... always.
The relevant method is write_type_matches in sampler_smirky.py. I will try to replicate the error.

davidlmobley commented:
@bannanc - I don't think you'd pushed (or maybe committed?) since the last time you updated from master, as your last commit brought in a bunch of changes with it.

bannanc commented:
I had it print the frac_found in the cli_smirky.py script to see if I could figure out the problem. 

0.577833125778 
that was printed, but then I asked for the type of frac_found and got
<type 'float'>

bannanc committed:
fixing how final percentage is printed

davidlmobley commented:
Typo above?

bannanc commented:
I think I fixed it I changed this line:
`if verbose: print("Final score was %.3f%%" % frac_found*100.)`
to 
`if verbose: print("Final score was %.3f %%" % (frac_found*100.0))`

davidlmobley commented:
@bannanc - that works for me now, but there still seems to be a print of the type left in your code that should be removed. 

Your Python2 test is also currently failing, but I don't see a good reason for it and the Python 3 tests are passing, so it seems possible it's some sort of sporadic failure. We'll see what happens with the next commit.

davidlmobley commented on examples/smirky/default_angle/README.md:
````diff
@@ -0,0 +1,38 @@
+
+From command line called:
+```
+smirky -molecules AlkEthOH_test_filt1_ff.mol2 --typetag Angle --verbose True
````
Should be `smirky --molecules`

davidlmobley commented on smarty/cli_smirky.py:
```diff
@@ -0,0 +1,178 @@
+            --iterations iterations --output outputFile]
+
+    example:
+    smirky -molecules AlkEthOH_test_filt1_ff.mol2 --typetag Angle
```
`smirky --molecules`

davidlmobley commented on smarty/cli_smirky.py:
```diff
@@ -0,0 +1,178 @@
+    parser.add_option('-a', '--bondANDdecors', metavar="DECORATORS",
+            action="store", type='string', dest='bond_AND_decorators',
+            default='odds_files/bond_AND_decorators.smarts',
+            help="Filename defining bond AND decorators and their associated probabilities. These are AND'd to the end of a bond, for example in '[#6]-,=;@[#7]' '@' is an AND decorator.(OPTIONAL)")
```
The help documentation is VERY WELL DONE for the CLI. Very nice.

davidlmobley commented on smarty/cli_smirky.py:
```diff
@@ -0,0 +1,178 @@
+    parser.add_option("-f", "--initialtypes", metavar='INITIALTYPES',
+            action="store", type="string", dest='initialtypes_filename',
+            default=None,
+            help="Filename defining initial (first) fragment types as 'SMIRKS    typename'. If this is left blank the initial type will be a generic form of the given fragment, for example '[*:1]~[*:2]' for a bond (OPTIONAL)")
```
I don't quite understand the help on this one. What's a "SMIRKS typename" and how is it relevant to the file here? (Maybe nonessential given the examples?)

bannanc replied:
All of the examples start with empty smirks, which is the default. 

The other option is to start with a set of initial SMIRKS, which have labels/typename which would correspond to the parameter ID's later when we read in a set of SMIRKS in a SMIRFF. 

I updated the wording a little so maybe it is more clear.

davidlmobley commented on smarty/cli_smirky.py:
```diff
@@ -0,0 +1,178 @@
+    parser.add_option('-r', '--smirff', metavar='REFERENCE',
+            action='store', type='string', dest='SMIRFF',
+            default=None,
+            help="Filename defining a SMIRFF force fielce used to determine reference fragment types in provided set of molecules. It may be an absolute file path, a path relative to the current working directory, or a path relative to this module's data subdirectory (for built in force fields). (OPTIONAL)")
```
"force fielce" -> "force field"

davidlmobley commented on smarty/data/odds_files/substitutions.smarts:
```diff
@@ -0,0 +1,9 @@
+% <SMARTS> <replacement-string>
+% halogens
+[#7!-1,#8,#16]              ewg2
+[#7!-1,#8!-1,#16!-1,$halogen]     ewg1
```
I had in my head that ewg1 and ewg2 were "different kinds" of electron withdrawing group in some sense but this makes them overlap. Correct?

bannanc replied:
These are copied from @cbayly13 
Traditionally: ewg1 are single bonded and ewg2 are double bonded. So oxygen, nitrogen, and sulfur are in both, but the halogens can only be single bonded.

bannanc committed:
remove testing print statements

bannanc commented:
I removed print statements, I'm running the tests on my computer to see if the python2 passes there.

bannanc commented:
The things failing here say "no module openeye.oechem" do I need to import openeye first for the Travis tool to work for some reason?

davidlmobley commented on smarty/environment.py:
```diff
@@ -128,27 +162,29 @@ def addANDtype(self, ANDtype):
                 added to the list of ANDtypes for this atom
             """
-            self.ANDtypes.append(ANDtype)
+            if ANDtype != "":
```
Should this also have `and ANDtype not in ANDtypes`?

bannanc replied:
yes, I updated this and updated a few other places to guarantee no repeated decorators

davidlmobley commented on smarty/environment.py:
```diff
@@ -226,90 +262,50 @@ def asSMIRKS(self):
-        def getORtypes(self):
-            """
-            returns a copy of the list of ORtypes for this bond
+        def getOrder(self):
```
Curious as I don't see this used anywhere -- what's the purpose of this function? (And why do "not triple" and "not aromatic" get bond orders of 1?)

bannanc replied:
We haven't put it into use, but Christopher and I wanted the ability to check valence around an atom as a possible way of proposing a move is valid. I'm not convinced it is necessary because we won't keep SMIRKS strings that don't type any molecule, but the methods exist if we want to be able to check them. 

The idea is that really you want the minimum bond order associated with a bond if you're going to filter by valence. for example if a bond is [_:1]-,=[_:2] the single or double bond should only count as 1 for bond order. The same idea goes for != or !: the minimum bond order for those is 1. the getOrder can then be used to get the minimum bond order around an atom in an environment.

davidlmobley replied:
OK. Maybe add a code comment that explains that this would be for valence checking? Otherwise anyone reading the code other than you is probably gonna wonder what is going on here and why the bond order is so wrong for "not triple", etc. I agree that this could be useful to have there and your explanation is sensible.

bannanc replied:
I added a line to the doc string, but it did say it returns the "minimum" bond order so 1 for "not" in front of triple, double, or aromatic is correct as far as the minimum bond that would meet that requirement.

davidlmobley commented on smarty/environment.py:
```diff
@@ -370,32 +366,64 @@ def _getAtomInfo(self, atom):
+            return None, []
+
+        # Save regular expression for SMIRKS decorators
+        reg = r'(!?[#]\d+|!?[aA]|!?[DHjrVX^]\d+|!?[R+-]\d*|!?[@]\d+|!?[@]@?)'
```
Please, no regexes without comments saying what they do... These are notoriously hard to read. Can you explain a pattern you're looking for in English or give example patterns?

bannanc replied:
Yes, I'll write in a comment for what this separates decorators.

bannanc replied:
I separate the reg into each decorator option with comments and then combined for the actual find all re function.

davidlmobley commented on smarty/environment.py:
```diff
@@ -465,70 +493,132 @@ def _asSMIRKS(self, initialAtom = None, neighbors = None, smarts = False):
+        descriptor: optional, None
+            None - returns any bond with equal probability
+            int - will return an bond with that index
+            'Indexed' - returns a random indexed bond
```
Do we have indexed bonds? I thought we only had indexed atoms? (Or does "indexed bond" mean a bond between indexed atoms?)

davidlmobley commented on smarty/sampler_smirky.py:
```diff
@@ -243,9 +314,37 @@ def __init__(self, molecules, typetag, elementList, ORdecorators,
+
     def get_force_type(self, typetag):
         """
         Uses typetag to get the Force type key word used to read the SMIRFF file
```
You have a comment in the code on line 366 asking what the force word is for impropers. Impropers are a `PeriodicTorsionForce` (they still exert a force via that mechanism) so they don't have their own keyword; they're distinguished only by what atoms the force is applied between.

bannanc replied:
Right, thanks, I forgot to go back in and update this.

davidlmobley commented on smarty/sampler_smirky.py:
```diff
@@ -683,21 +904,26 @@ def sample_types(self):
                     return False
 
             # updated proposed parent dictionary
             # TODO: update parent dictionary
```
You have "TODO" items strewn a few places through this code. Maybe doublecheck occurrences of the word and make sure none should be resolved right now, and then create an issue to resolve the others (tagged as low priority if it is indeed low priority)...

bannanc replied:
Made issues #140 and #141

davidlmobley reviewed (approved):
I've pointed out a number of minor issues above -- mostly documentation related -- that you should review; some should probably be fixed. After that I think this is good to merge. Thanks, @bannanc .

davidlmobley commented:
@bannanc -- it looks like the first error in your Travis log for python 2 is this:

```
  Could not find a version that satisfies the requirement OpenEye-toolkits-python2.7-ubuntu-12.04-x64==2016.10b3 (from openeye-toolkits) (from versions: 2016.10b1, 2016.10b2)
No matching distribution found for OpenEye-toolkits-python2.7-ubuntu-12.04-x64==2016.10b3 (from openeye-toolkits)
```

It looks to me like they just came out with a new OpenEye beta version (10b3) which is not yet available for the python 2.7 ubuntu12 platform, so Travis is failing to install the openeye toolkits which results in everything else failing.

This is OK, actually. We can merge your PR when it is ready to merge (assuming python 2.7 passes locally) because:
a) it's a problem which should go away as soon as the 10b3 is up for ubuntu 12
b) it works locally
c) the other python versions pass
etc.

davidlmobley commented:
@bannanc : 👍  on all the comments you just made. Sounds like you're going to be pretty close (once you make some updates to the overall README.md) to having a merge-able version.

bannanc committed:
minor changes in code and README example

bannanc committed:
updated README with smirky documentation

bannanc commented:
@davidlmobley I think this is updated, I added the help documentation from smirky to the primary README file. I ran the tests again locally to double check and they are still passing on python2.7

The smarty part of the README seems a little out of date, it includes decorators I don't think we're using any more, Camila would be more qualified to comment on that, maybe she can update it when she gets her next pull request ready.

davidlmobley commented:
This looks fine to me, @bannanc . The one thing I'd probably add (but it's not necessary at this point if you're in a rush to get this off your plate) is to explain in the README how the scoring is done, and that it is the analog of SMARTY but extended to these other "types" or environments.

There's also a minor typo on one of the new lines the README (just a jumbled up word).

I'm stepping out for a minute, but ping me after you see this, when the tests pass, or at least when the Python 3 tests pass, and I can merge.

bannanc committed:
minor README changes

bannanc commented:
slight updates on the README.

bannanc committed:
getOrder doc string update
